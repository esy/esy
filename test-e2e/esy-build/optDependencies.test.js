// @flow

const helpers = require('../test/helpers');
const {test, isWindows, isMacos, isLinux} = helpers;

function makePackage(p, {name, build, dependencies, optDependencies}, ...children) {
  return [
    helpers.packageJson({
      name: name,
      version: '1.0.0',
      esy: {
        build,
      },
      dependencies,
      optDependencies,
    }),
    ...children,
  ];
}

describe('Build with optDependencies', () => {
  describe('builds w/o opt dependency installed', () => {
    async function createTestSandbox() {
      const p = await helpers.createTestSandbox();

      await p.fixture(
        ...makePackage(
          p,
          {
            name: 'root',
            build: 'true',
            dependencies: {
              dep: 'path:./dep',
            },
            optDependencies: {},
          },
          helpers.dir(
            'dep',
            ...makePackage(p, {
              name: 'dep',
              build: 'optDep.installed #{optDep.installed}',
              dependencies: {},
              optDependencies: {optDep: '*'},
            }),
          ),
        ),
      );
      return p;
    }

    test('check build-plan', async () => {
      const p = await createTestSandbox();
      await p.esy('install');
      const plan = JSON.parse((await p.esy('build-plan -p dep@path:dep')).stdout);
      expect(plan.build).toEqual([['optDep.installed', 'false']]);
    });

    test.enableIf(isMacos || isLinux)('snapshot build-env', async () => {
      const p = await createTestSandbox();
      await p.esy('install');

      {
        const id = JSON.parse((await p.esy('build-plan')).stdout).id;
        const depid = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const {stdout} = await p.esy('build-env --build-concurrency 16');
        expect(p.normalizePathsForSnapshot(stdout, {id, depid})).toMatchSnapshot();
      }

      {
        const id = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const {stdout} = await p.esy('build-env -p dep --build-concurrency 16');
        expect(p.normalizePathsForSnapshot(stdout, {id})).toMatchSnapshot();
      }
    });
  });

  describe('builds w/ opt dependency installed', () => {
    async function createTestSandbox() {
      const p = await helpers.createTestSandbox();

      await p.fixture(
        ...makePackage(
          p,
          {
            name: 'root',
            build: 'true',
            dependencies: {
              dep: 'path:./dep',
              optDep: 'path:./optDep',
            },
            optDependencies: {},
          },
          helpers.dir(
            'dep',
            ...makePackage(p, {
              name: 'dep',
              build: 'optDep.installed #{optDep.installed}',
              dependencies: {},
              optDependencies: {optDep: '*'},
            }),
          ),
          helpers.dir(
            'optDep',
            ...makePackage(p, {
              name: 'optDep',
              build: 'true',
              dependencies: {},
              optDependencies: {},
            }),
          ),
        ),
      );

      return p;
    }

    test('check build-plan', async () => {
      const p = await createTestSandbox();
      await p.esy('install');
      const plan = JSON.parse((await p.esy('build-plan -p dep@path:dep')).stdout);
      expect(plan.build).toEqual([['optDep.installed', 'true']]);
    });

    test.enableIf(isMacos || isLinux)('snapshot build-env', async () => {
      const p = await createTestSandbox();
      await p.esy('install');
      {
        const id = JSON.parse((await p.esy('build-plan')).stdout).id;
        const depid = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const optdepid = JSON.parse((await p.esy('build-plan -p optDep')).stdout).id;
        const {stdout} = await p.esy('build-env --build-concurrency 16');
        expect(
          p.normalizePathsForSnapshot(stdout, {id, depid, optdepid}),
        ).toMatchSnapshot();
      }

      {
        const id = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const optdepid = JSON.parse((await p.esy('build-plan -p optDep')).stdout).id;
        const {stdout} = await p.esy('build-env -p dep --build-concurrency 16');
        expect(p.normalizePathsForSnapshot(stdout, {id, optdepid})).toMatchSnapshot();
      }
    });
  });

  test('opam package builds w/ opt dependency installed', async () => {
    const p = await helpers.createTestSandbox();

    await p.defineNpmPackage({
      name: '@esy-ocaml/substs',
      version: '1.0.0',
      esy: {},
    });

    await p.defineOpamPackage({
      name: 'dep',
      version: '1.0.0',
      opam: `
      opam-version: "2.0"
      build: [
        ["optDep:installed" "%{optDep:installed}%"]
        ["optDep:enabled" "%{optDep:enabled}%"]
      ]
      depopts: ["optDep"]
      `,
      url: null,
    });

    await p.defineOpamPackage({
      name: 'optDep',
      version: '1.0.0',
      opam: `
      opam-version: "2.0"
      build: [
        "true"
      ]
      `,
      url: null,
    });

    await p.fixture(
      ...makePackage(p, {
        name: 'root',
        build: 'true',
        dependencies: {
          '@opam/dep': '*',
          '@opam/optDep': '*',
        },
        optDependencies: {},
      }),
    );

    await p.esy('install');
    const plan = JSON.parse((await p.esy('build-plan -p @opam/dep@opam:1.0.0')).stdout);
    expect(plan.build).toEqual([
      ['optDep:installed', 'true'],
      ['optDep:enabled', 'enable'],
    ]);
  });

  test('opam package builds w/o opt dependency installed', async () => {
    const p = await helpers.createTestSandbox();

    await p.defineNpmPackage({
      name: '@esy-ocaml/substs',
      version: '1.0.0',
      esy: {},
    });

    await p.defineOpamPackage({
      name: 'dep',
      version: '1.0.0',
      opam: `
      opam-version: "2.0"
      build: [
        ["optDep:installed" "%{optDep:installed}%"]
        ["optDep:enabled" "%{optDep:enabled}%"]
      ]
      depopts: ["optDep"]
      `,
      url: null,
    });

    await p.defineOpamPackage({
      name: 'optDep',
      version: '1.0.0',
      opam: `
      opam-version: "2.0"
      build: [
        "true"
      ]
      `,
      url: null,
    });

    await p.fixture(
      ...makePackage(p, {
        name: 'root',
        build: 'true',
        dependencies: {
          '@opam/dep': '*',
        },
        optDependencies: {},
      }),
    );

    await p.esy('install');
    const plan = JSON.parse((await p.esy('build-plan -p @opam/dep@opam:1.0.0')).stdout);
    expect(plan.build).toEqual([
      ['optDep:installed', 'false'],
      ['optDep:enabled', 'disable'],
    ]);
  });
});
