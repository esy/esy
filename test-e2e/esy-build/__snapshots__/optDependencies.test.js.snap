// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build with optDependencies builds w/ opt dependency installed snapshot build-env 1`] = `
"# Build environment
# package:            root@link-dev:./package.json
# depspec:            dependencies(self)+devDependencies(self)
# mode:               buildDev
# envspec:            
# buildIsInProgress:  true
# includeBuildEnv:    true
# includeCurrentEnv:  false
# includeNpmBin:      false


#
# Built-in
#
export cur__jobs="8"
export cur__etc="%projectPath%/_esy/default/store/i/%id%/etc"
export cur__share="%projectPath%/_esy/default/store/i/%id%/share"
export cur__toplevel="%projectPath%/_esy/default/store/i/%id%/toplevel"
export cur__stublibs="%projectPath%/_esy/default/store/i/%id%/stublibs"
export cur__doc="%projectPath%/_esy/default/store/i/%id%/doc"
export cur__man="%projectPath%/_esy/default/store/i/%id%/man"
export cur__lib="%projectPath%/_esy/default/store/i/%id%/lib"
export cur__sbin="%projectPath%/_esy/default/store/i/%id%/sbin"
export cur__bin="%projectPath%/_esy/default/store/i/%id%/bin"
export cur__install="%projectPath%/_esy/default/store/i/%id%"
export cur__target_dir="%projectPath%/_esy/default/store/b/%id%"
export cur__original_root="%projectPath%"
export cur__root="%projectPath%"
export cur__dev="true"
export cur__version="1.0.0"
export cur__name="root"

#
# optDep@path:optDep@d41d8cd9
#
export CAML_LD_LIBRARY_PATH="%esyStorePath%/i/%optdepid%/stublibs:%esyStorePath%/i/%optdepid%/lib/stublibs:$CAML_LD_LIBRARY_PATH"
export OCAMLPATH="%esyStorePath%/i/%optdepid%/lib:$OCAMLPATH"
export MAN_PATH="%esyStorePath%/i/%optdepid%/man:$MAN_PATH"
export PATH="%esyStorePath%/i/%optdepid%/bin:$PATH"

#
# dep@path:dep@d41d8cd9
#
export CAML_LD_LIBRARY_PATH="%esyStorePath%/i/%depid%/stublibs:%esyStorePath%/i/%depid%/lib/stublibs:$CAML_LD_LIBRARY_PATH"
export OCAMLPATH="%esyStorePath%/i/%depid%/lib:$OCAMLPATH"
export MAN_PATH="%esyStorePath%/i/%depid%/man:$MAN_PATH"
export PATH="%esyStorePath%/i/%depid%/bin:$PATH"

#
# Built-in
#
export OCAMLFIND_LDCONF="ignore"
export OCAMLFIND_DESTDIR="%projectPath%/_esy/default/store/i/%id%/lib"
export DUNE_BUILD_DIR="%projectPath%/_esy/default/store/b/%id%"
export DUNE_STORE_ORIG_SOURCE_DIR="true"
export SHELL="env -i /bin/bash --norc --noprofile"
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin""
`;

exports[`Build with optDependencies builds w/ opt dependency installed snapshot build-env 2`] = `
"# Build environment
# package:            dep@path:dep
# depspec:            dependencies(self)
# mode:               build
# envspec:            
# buildIsInProgress:  true
# includeBuildEnv:    true
# includeCurrentEnv:  false
# includeNpmBin:      false


#
# Built-in
#
export cur__jobs="8"
export cur__etc="%esyStorePath%/s/%id%/etc"
export cur__share="%esyStorePath%/s/%id%/share"
export cur__toplevel="%esyStorePath%/s/%id%/toplevel"
export cur__stublibs="%esyStorePath%/s/%id%/stublibs"
export cur__doc="%esyStorePath%/s/%id%/doc"
export cur__man="%esyStorePath%/s/%id%/man"
export cur__lib="%esyStorePath%/s/%id%/lib"
export cur__sbin="%esyStorePath%/s/%id%/sbin"
export cur__bin="%esyStorePath%/s/%id%/bin"
export cur__install="%esyStorePath%/s/%id%"
export cur__target_dir="%esyPrefixPath%/3/b/%id%"
export cur__original_root="%esyPrefixPath%/source/i/dep__5d502355"
export cur__root="%esyPrefixPath%/source/i/dep__5d502355"
export cur__dev="false"
export cur__version="1.0.0"
export cur__name="dep"

#
# optDep@path:optDep@d41d8cd9
#
export CAML_LD_LIBRARY_PATH="%esyStorePath%/i/%optdepid%/stublibs:%esyStorePath%/i/%optdepid%/lib/stublibs:$CAML_LD_LIBRARY_PATH"
export OCAMLPATH="%esyStorePath%/i/%optdepid%/lib:$OCAMLPATH"
export MAN_PATH="%esyStorePath%/i/%optdepid%/man:$MAN_PATH"
export PATH="%esyStorePath%/i/%optdepid%/bin:$PATH"

#
# Built-in
#
export OCAMLFIND_LDCONF="ignore"
export OCAMLFIND_DESTDIR="%esyStorePath%/s/%id%/lib"
export DUNE_BUILD_DIR="%esyPrefixPath%/3/b/%id%"
export SHELL="env -i /bin/bash --norc --noprofile"
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin""
`;

exports[`Build with optDependencies builds w/o opt dependency installed snapshot build-env 1`] = `
"# Build environment
# package:            root@link-dev:./package.json
# depspec:            dependencies(self)+devDependencies(self)
# mode:               buildDev
# envspec:            
# buildIsInProgress:  true
# includeBuildEnv:    true
# includeCurrentEnv:  false
# includeNpmBin:      false


#
# Built-in
#
export cur__jobs="8"
export cur__etc="%projectPath%/_esy/default/store/i/%id%/etc"
export cur__share="%projectPath%/_esy/default/store/i/%id%/share"
export cur__toplevel="%projectPath%/_esy/default/store/i/%id%/toplevel"
export cur__stublibs="%projectPath%/_esy/default/store/i/%id%/stublibs"
export cur__doc="%projectPath%/_esy/default/store/i/%id%/doc"
export cur__man="%projectPath%/_esy/default/store/i/%id%/man"
export cur__lib="%projectPath%/_esy/default/store/i/%id%/lib"
export cur__sbin="%projectPath%/_esy/default/store/i/%id%/sbin"
export cur__bin="%projectPath%/_esy/default/store/i/%id%/bin"
export cur__install="%projectPath%/_esy/default/store/i/%id%"
export cur__target_dir="%projectPath%/_esy/default/store/b/%id%"
export cur__original_root="%projectPath%"
export cur__root="%projectPath%"
export cur__dev="true"
export cur__version="1.0.0"
export cur__name="root"

#
# dep@path:dep@d41d8cd9
#
export CAML_LD_LIBRARY_PATH="%esyStorePath%/i/%depid%/stublibs:%esyStorePath%/i/%depid%/lib/stublibs:$CAML_LD_LIBRARY_PATH"
export OCAMLPATH="%esyStorePath%/i/%depid%/lib:$OCAMLPATH"
export MAN_PATH="%esyStorePath%/i/%depid%/man:$MAN_PATH"
export PATH="%esyStorePath%/i/%depid%/bin:$PATH"

#
# Built-in
#
export OCAMLFIND_LDCONF="ignore"
export OCAMLFIND_DESTDIR="%projectPath%/_esy/default/store/i/%id%/lib"
export DUNE_BUILD_DIR="%projectPath%/_esy/default/store/b/%id%"
export DUNE_STORE_ORIG_SOURCE_DIR="true"
export SHELL="env -i /bin/bash --norc --noprofile"
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin""
`;

exports[`Build with optDependencies builds w/o opt dependency installed snapshot build-env 2`] = `
"# Build environment
# package:            dep@path:dep
# depspec:            dependencies(self)
# mode:               build
# envspec:            
# buildIsInProgress:  true
# includeBuildEnv:    true
# includeCurrentEnv:  false
# includeNpmBin:      false


#
# Built-in
#
export cur__jobs="8"
export cur__etc="%esyStorePath%/s/%id%/etc"
export cur__share="%esyStorePath%/s/%id%/share"
export cur__toplevel="%esyStorePath%/s/%id%/toplevel"
export cur__stublibs="%esyStorePath%/s/%id%/stublibs"
export cur__doc="%esyStorePath%/s/%id%/doc"
export cur__man="%esyStorePath%/s/%id%/man"
export cur__lib="%esyStorePath%/s/%id%/lib"
export cur__sbin="%esyStorePath%/s/%id%/sbin"
export cur__bin="%esyStorePath%/s/%id%/bin"
export cur__install="%esyStorePath%/s/%id%"
export cur__target_dir="%esyPrefixPath%/3/b/%id%"
export cur__original_root="%esyPrefixPath%/source/i/dep__5d502355"
export cur__root="%esyPrefixPath%/source/i/dep__5d502355"
export cur__dev="false"
export cur__version="1.0.0"
export cur__name="dep"
export OCAMLFIND_LDCONF="ignore"
export OCAMLFIND_DESTDIR="%esyStorePath%/s/%id%/lib"
export DUNE_BUILD_DIR="%esyPrefixPath%/3/b/%id%"
export SHELL="env -i /bin/bash --norc --noprofile"
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin""
`;
